kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/JvmDescriptorUtil.kt

```diff
@@ -0,0 +1,60 @@
+package com.squareup.moshi.kotlin.reflect
+
+import kotlinx.metadata.jvm.JvmFieldSignature
+import kotlinx.metadata.jvm.JvmMethodSignature
+import java.lang.reflect.Constructor
+import java.lang.reflect.Field
+import java.lang.reflect.Method
+
+internal val Class<*>.descriptor: String
+  get() {
+    return when {
+      isPrimitive -> when (kotlin) {
+        Byte::class -> "B"
+        Char::class -> "C"
+        Double::class -> "D"
+        Float::class -> "F"
+        Int::class -> "I"
+        Long::class -> "J"
+        Short::class -> "S"
+        Boolean::class -> "Z"
+        Void::class -> "V"
+        else -> throw RuntimeException("Unrecognized primitive $this")
+      }
+      isArray -> name.replace('.', '/')
+      else -> "L$name;".replace('.', '/')
+    }
+  }
+
+internal val Method.descriptor: String
+  get() = parameterTypes.joinToString(separator = "", prefix = "(", postfix = ")${returnType.descriptor}") { it.descriptor }
+
+/**
+ * Returns the JVM signature in the form "$Name$MethodDescriptor", for example: `equals(Ljava/lang/Object;)Z`.
+ *
+ * Useful for comparing with [JvmMethodSignature].
+ *
+ * For reference, see the [JVM specification, section 4.3](http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3).
+ */
+internal val Method.jvmMethodSignature: String get() = "$name$descriptor"
+
+internal val Constructor<*>.descriptor: String
```

Comment:
do we use both `descriptor` and `jvmMethodSignature` forms? Otherwise would be nice to only have one!

Reply:
5af57b6